### PR TITLE
[WIP] list cmd

### PR DESCRIPTION
**Pull request by kerlyn-bsd3**
This is the initial PR for the 'list' function. Support has been added for gdb-style *linespecs*, which can be used as a parameter to several functions that take 'location' (e.g. break, clear, until, list). List currently works within the current source file and supported forms are 'list', 'list -', and 'list *linenum*'. There is one known bug, which allows the internal list context to run off the end of the file. The workaround is to use 'list -'.

### CODE REVIEW

**Comment by derekparker** [comment]
Hey, sorry for the delay in reviewing this, I'll have feedback for you tomorrow after I get a chance to look through everything.

**Comment by kerlyn-bsd3** [comment]
@derekparker 
NP - I'm sure you'll have some questions and I'm happy to discuss.

File: proctl/proctl.go [hunk 1]
```diff
@@ -45,6 +46,21 @@ type DebuggedProcess struct {
 	exited                  bool
 }
 
+// List struct mirrors api.List{}
```
Comment (derekparker) [inline proctl/proctl.go]: Omit {}

File: proctl/proctl.go [hunk 2]
```diff
@@ -45,6 +46,21 @@ type DebuggedProcess struct {
...
+type List struct {
+	// File & line information for printing.
```
Comment (derekparker) [inline proctl/proctl.go]: Drop this comment.

File: proctl/proctl.go [hunk 3]
```diff
@@ -45,6 +46,21 @@ type DebuggedProcess struct {
...
+	Addr         uint64 // Address breakpoint is set for.
```
Comment (derekparker) [inline proctl/proctl.go]: This structure is not unique to breakpoints.

File: proctl/proctl.go [hunk 4]
```diff
@@ -114,9 +131,12 @@ func (dbp *DebuggedProcess) LoadInformation(path string) error {
 	return nil
 }
 
-// Find a location by string (file+line, function, breakpoint id, addr)
+// Find a location by string (function, linenum, filename:linenum, *address)
 func (dbp *DebuggedProcess) FindLocation(str string) (uint64, error) {
-	// File + Line
+
```
Comment (derekparker) [inline proctl/proctl.go]: Kill whitespace.

File: proctl/proctl.go [hunk 5]
```diff
@@ -143,13 +192,19 @@ func (dbp *DebuggedProcess) FindLocation(str string) (uint64, error) {
 		return fn.Entry, nil
 	}
 
-	// Attempt to parse as number for breakpoint id or raw address
+	// No joy
+	return 0, nil
```
Comment (derekparker) [inline proctl/proctl.go]: This should really return an error when no location is actually found.

File: proctl/proctl.go [hunk 6]
```diff
@@ -143,13 +192,19 @@ func (dbp *DebuggedProcess) FindLocation(str string) (uint64, error) {
...
+}
+
+// Find a breakpoint by string (breakpoint id)
+func (dbp *DebuggedProcess) FindBp(str string) (uint64, error) {
```
Comment (derekparker) [inline proctl/proctl.go]: This should probably just be inlined.

File: proctl/proctl.go [hunk 7]
```diff
@@ -143,13 +192,19 @@ func (dbp *DebuggedProcess) FindLocation(str string) (uint64, error) {
...
+
+	// Attempt to parse number to use as breakpoint id
 	id, err := strconv.ParseUint(str, 0, 64)
 	if err != nil {
-		return 0, fmt.Errorf("unable to find location for %s", str)
+		return 0, fmt.Errorf("unable to find bp %s", str)
```
Comment (derekparker) [inline proctl/proctl.go]: s/bp/breakpoint/

Also, this would actually be clearer as "invalid breakpoint id %s".

File: proctl/proctl.go [hunk 8]
```diff
@@ -225,6 +279,15 @@ func (dbp *DebuggedProcess) ClearByLocation(loc string) (*BreakPoint, error) {
 	return dbp.Clear(addr)
 }
 
+// Deletes a breakpoint by breakpoint id
+func (dbp *DebuggedProcess) DeleteById(id string) (*BreakPoint, error) {
```
Comment (derekparker) [inline proctl/proctl.go]: This function does not seem to be used.

File: proctl/stack.go [hunk 9]
```diff
@@ -13,7 +13,7 @@ type stackLocation struct {
 }
 
 // Takes an offset from RSP and returns the address of the
-// instruction the currect function is going to return to.
+// instruction the current function is going to return to.
```
Comment (derekparker) [inline proctl/stack.go]: Thanks, good catch!

File: service/api/types.go [hunk 10]
```diff
@@ -2,45 +2,72 @@ package api
...
 	// CurrentThread is the currently selected debugger thread.
 	CurrentThread *Thread `json:"currentThread,omitempty"`
+	// List is the current focus for printing source lines.
+	List *List `json:"list,omitempty"`
 	// Exited indicates whether the debugged process has exited.
 	Exited bool `json:"exited"`
 }
 
+// Arguments holds the unparsed command line argument(s).
+type Arguments struct {
```
Comment (derekparker) [inline service/api/types.go]: This name seems to general, and I don't think it buys us anything over just using a slice.

File: service/client.go [hunk 11]
```diff
@@ -27,7 +27,7 @@ type Client interface {
 	// GetBreakPoint gets a breakpoint by ID.
 	GetBreakPoint(id int) (*api.BreakPoint, error)
 	// CreateBreakPoint creates a new breakpoint.
-	CreateBreakPoint(*api.BreakPoint) (*api.BreakPoint, error)
+	CreateBreakPoint(location *api.Arguments) (*api.BreakPoint, error)
```
Comment (derekparker) [inline service/client.go]: Drop 'location'.

File: service/debugger/debugger.go [hunk 12]
```diff
@@ -184,9 +185,29 @@ func (d *Debugger) State() (*api.DebuggerState, error) {
 			breakpoint = convertBreakPoint(bp)
 		}
 
+		lc := p.ListContext
+		if lc == nil {
+			log.Println("*** Nil list context ***")
+			list := new(proctl.List)
```
Comment (derekparker) [inline service/debugger/debugger.go]: Should just be `lc = &proctl.ListP`{...}`

File: service/debugger/debugger.go [hunk 13]
```diff
@@ -195,19 +216,14 @@ func (d *Debugger) State() (*api.DebuggerState, error) {
...
-		case len(requestedBp.File) > 0:
-			loc = fmt.Sprintf("%s:%d", requestedBp.File, requestedBp.Line)
-		case len(requestedBp.FunctionName) > 0:
-			loc = requestedBp.FunctionName
-		default:
-			return fmt.Errorf("no file or function name specified")
-		}
+		var loc string = ""
 
+		if len(location.Args) == 1 {
```
Comment (derekparker) [inline service/debugger/debugger.go]: What if it doesn't?

File: service/debugger/debugger.go [hunk 14]
```diff
@@ -328,6 +344,56 @@ func (d *Debugger) Command(command *api.DebuggerCommand) (*api.DebuggerState, er
...
 }
 
+func (d *Debugger) ListSource(location string) (*api.DebuggerState, error) {
+	state, err := d.State()
+	if state.List == nil {
+		return nil, nil
+	}
+
+	err = d.withProcess(func(p *proctl.DebuggedProcess) error {
+
```
Comment (derekparker) [inline service/debugger/debugger.go]: Kill whitespace.

File: service/debugger/debugger.go [hunk 15]
```diff
@@ -328,6 +344,56 @@ func (d *Debugger) Command(command *api.DebuggerCommand) (*api.DebuggerState, er
...
+		if state.List.Line == 0 {
+			state.List.File = state.CurrentThread.File
+			state.List.Line = state.CurrentThread.Line
+		}
+		state.List.Direction = true
+		numLines := state.List.ListSize
+		halfLines := (numLines + 1) / 2
+
+		if len(location) != 0 {
```
Comment (derekparker) [inline service/debugger/debugger.go]: This would be clearer as `location != ""`

File: terminal/command.go [hunk 16]
```diff
@@ -40,25 +40,28 @@ type Commands struct {
 	client  service.Client
 }
 
+var c *Commands
```
Comment (derekparker) [inline terminal/command.go]: Why promote this out of the function?

File: terminal/command.go [hunk 17]
```diff
@@ -40,25 +40,28 @@ type Commands struct {
...
+
 // Returns a Commands struct with default commands defined.
 func DebugCommands(client service.Client) *Commands {
-	c := &Commands{client: client}
+	c = &Commands{client: client}
 
 	c.cmds = []command{
```
Comment (derekparker) [inline terminal/command.go]: Could you break this aliases out into a separate PR?

File: terminal/command.go [hunk 18]
```diff
@@ -40,25 +40,28 @@ type Commands struct {
...
-		{aliases: []string{"help"}, cmdFn: c.help, helpMsg: "Prints the help message."},
+		{aliases: []string{"help", "h"}, cmdFn: c.help, helpMsg: "Prints the help message."},
 		{aliases: []string{"break", "b"}, cmdFn: breakpoint, helpMsg: "Set break point at the entry point of a function, or at a specific file/line. Example: break foo.go:13"},
 		{aliases: []string{"continue", "c"}, cmdFn: cont, helpMsg: "Run until breakpoint or program termination."},
 		{aliases: []string{"step", "si"}, cmdFn: step, helpMsg: "Single step through program."},
 		{aliases: []string{"next", "n"}, cmdFn: next, helpMsg: "Step over to next source line."},
-		{aliases: []string{"threads"}, cmdFn: threads, helpMsg: "Print out info for every traced thread."},
+		{aliases: []string{"threads", "allm"}, cmdFn: threads, helpMsg: "Print out info for every traced thread."},
```
Comment (derekparker) [inline terminal/command.go]: Not necessarily on board with this alias, a casual user would likely not know that internally the scheduler refers to threads as Ms.

File: terminal/command.go [hunk 19]
```diff
@@ -78,7 +81,7 @@ func (c *Commands) Register(cmdstr string, cf cmdfunc, helpMsg string) {
 }
 
 // Find will look up the command function for the given command input.
-// If it cannot find the command it will defualt to noCmdAvailable().
+// If it cannot find the command it will default to noCmdAvailable().
```
Comment (derekparker) [inline terminal/command.go]: Thanks!

File: terminal/command.go [hunk 20]
```diff
@@ -408,18 +421,37 @@ func printcontext(state *api.DebuggerState) error {
...
+	if state.List.Line < halfLines {
+		state.List.Line = halfLines
+	}
+	state.List.FirstLine = state.List.Line - halfLines + 1
+	state.List.LastLine = state.List.Line + halfLines
+
+	return print(state)
+}
+
+func print(state *api.DebuggerState) error {
```
Comment (derekparker) [inline terminal/command.go]: I think it'd be great to have this be a method off of `Line` instead of as a standalone function that returns a formatted string.

**Comment by derekparker** [comment]
Would you mind breaking out spelling corrections / new aliases out into separate PRs?

**Comment by derekparker** [comment]
@kerlyn-bsd3 any chance you'd have some time to go through the feedback this weekend?

**Comment by kerlyn-bsd3** [comment]
Yes, sorry.  I've been down a rabbit hole at work...

**Comment by derekparker** [comment]
@kerlyn-bsd3 just a heads up, I've done some refactoring / organizing that could impact this.

I really think this PR could be split into multiple:

- Fix typos
- Refactor: Implement linespec for breakpoints (setting, clearing)
- Add List cmd

I figure you've been busy, but just wanted to give a heads up. Also, if you need any help with this commit, let me know. If you feel like you won't have the time to rebase and push it through please let me know, as I can take it on and give you credit for the great work you've done.

**Comment by derekparker** [comment]
Going to close this PR due to inactivity and recent discussions happening in https://groups.google.com/forum/#!topic/delve-dev/HLwyfB1kYT4
